src/runtime_src/core/pcie/tools/xbmgmt/cmd_flash.cpp

```diff
@@ -142,45 +157,75 @@ static int updateSC(unsigned index, const char *file)
     if(!flasher.isValid())
         return -EINVAL;
 
-    auto dev = pcidev::get_dev(index, true);
+    auto mgmt_dev = pcidev::get_dev(index, false);
+    std::shared_ptr<pcidev::pci_device> dev;
+    int i = 0;
+    for (dev = pcidev::get_dev(i, true); dev; dev = pcidev::get_dev(i, true)) {
+        if(dev->domain == mgmt_dev->domain &&
+            dev->bus == mgmt_dev->bus &&
+	    dev->dev == mgmt_dev->dev) {
+                break;
+        }
+	i++;
+    }
+
+    if (!dev) {
+        return writeSCImage(flasher, file);
+    }
 
-    if (dev->sysfs_name.empty()) {
+    std::string errmsg;
+    std::string tmp;
+    dev->sysfs_get("", "user_pf", errmsg, tmp);
+    if (!errmsg.empty()) {
         std::cout << "CAUTION: User function is not found. " <<
-            "This is probably due to user function is running in virtual machine. " << std::endl;
+            "This is probably due to user function is running in virtual machine or user driver is not loaded. " << std::endl;
         if(!canProceed())
             return -ECANCELED;
+	return writeSCImage(flasher, file);
     }
 
-    std::string errmsg;
     std::cout << "Stopping user function..." << std::endl;
-    dev->sysfs_put("", "remove", errmsg, "1\n");
+
+    dev->sysfs_put("", "shutdown", errmsg, "1\n");
     if (!errmsg.empty()) {
-        std::cout << "Stopping user function failed" << std::endl;
+        std::cout << "Shutdown user function failed." << std::endl;
 	return -EINVAL;
     }
 
-    const std::string rescan_path = "/sys/bus/pci/rescan";
-    std::fstream fs;
-    std::ios::openmode mode = std::ios::out;
+    /* Poll till shutdown is done */
+    int shutdownStatus = 0;
+    for (int wait = 0; wait < SHUTDOWN_TIMEOUT; wait++) {
+        dev->sysfs_get<int>("", "shutdown", errmsg, shutdownStatus, EINVAL);
+        if (!errmsg.empty()) {
+            std::cout << errmsg << std::endl;
+            return -EINVAL;
+        }
 
-    fs.open(rescan_path, mode);
-    if (!fs.is_open()) {
-        perror(rescan_path.c_str());
-        return -EINVAL;
+        if (shutdownStatus == 1){
+            /* Shutdown is done successfully. Returning from here */
+            break;
+        }
+        sleep(1);
     }
 
+    if (!shutdownStatus) {
+        std::cout << "Shutdown user function timeout." << std::endl;
+	return -ETIMEDOUT;
+    }
 
-
-    std::shared_ptr<firmwareImage> bmc =
-        std::make_shared<firmwareImage>(file, BMC_FIRMWARE);
-    if (bmc->fail()) {
-        ret = -EINVAL;
-    } else {
-        ret = flasher.upgradeBMCFirmware(bmc.get());
+    dev->sysfs_put("", "remove", errmsg, "1\n");
+    if (!errmsg.empty()) {
+        std::cout << "Stopping user function failed" << std::endl;
+	return -EINVAL;
     }
 
-    fs << "1\n";
-    fs.flush();
+    sleep(15);
+    ret = writeSCImage(flasher, file);
+
+    mgmt_dev->sysfs_put("", "dparent/rescan", errmsg, "1\n");
+    if (!errmsg.empty()) {
+        std::cout << "ERROR: " << errmsg << std::endl;
+    }
```

Comment:
should we wait for user pf to show up again before return?

Reply:
done